### DISCUSSION TITLE
Really immutable collections

### DISCUSSION THREAD

**Discussion by golaod** [laravel/framework#53050] [Ideas]
You can read in the documentation:
> In general, collections are immutable, meaning every Collection method returns "an entirely" new Collection instance.

But that statemen is not true, because default collection implements `\ArrayAccess` interface. How about creating AbstractCollection (or trait) that holds most of the logic and have 2 child classes to represent mutable and immutable collections. For backward compatibility the name for mutable collection would stay the same and it would implement `\ArrayAccess` whereas `ImmutableCollection` would have `add/remove` would simply use `merge/except` under the hood.

**Comment by henzeb** [comment]
Collection already implements EnumeratesValues. I suggest 2 simple classes: ImmutableCollection and LazyImmutableCollection. the add/remove methods would do merge/except, and the magic offsetSet /offsetUnset methods would throw exceptions. 

you could easily do that yourself by the way.

**Reply by golaod** [reply]
That's not clean and that's why I proposed to move most of the code to AbstractCollection or Trait, so that it can be properly reused in Collection and ImmutableCollection.

**Reply by henzeb** [reply]
Why is that not clean? 

most of the code is already in trait EnumeratesValues, so moving it there would be the same as AbstractCollection

**Reply by golaod** [reply]
It's not clean to throw exceptions, because given class doesn't implement them. If it doesn't then it shouldn't be part of the parent then. Proper abstraction doesn't require to throw exception about invalid operations. I understand that there might be cases where it's true, but probably even those can be replaced with other design pattern like visitor.

The collection class literally what it should add to basic logic is the data access and any other method that is actually mutating items, whereas abstract would hold everything else that can be shared between Collection and ImmutableCollection. You prefer trait, fine but it still better to add methods than to throw exception cause you don't want them in your class.

**Reply by henzeb** [reply]
I still prefer ImmutableCollection extends Collection rather than two different objects that can't be used without defining union types.

 Returning a new instance like most methods do already should be a logical choice then...

**Reply by golaod** [reply]
Edit because I couldn't understand myself:
But I proposed AbstractCollection that would hold logic from both classes and only Collection would implement methods related to array access.
There would be also no issue with union types, because either you can use AbstractCollection for typehints OR we could add collection interface.
